Create ColumnRace.py

AniekpongEffiong opened the pull request:
This PR introduces a new column_race chart class for creating animated bar chart race visualizations, designed to work seamlessly with the SJVisualizer framework.

Key Features:
1. Animated Column Transitions: Smoothly animates bar height changes and positional movements as data rankings evolve over time.
2. Top-N Ranking Display: Automatically sorts and displays the top N data entries at each frame.
3. Dual Image Support: Internal column images (positioned inside bars) & External floating images (positioned above bars).
4. Negative Value Handling: Supports negative values with distinct red coloring and inverted bar direction.
5. Customizable Axes: Integrated with ColumnAxis.py for professional gridlines and scaling.
6. Dynamic Data Updates: Efficiently adds/removes columns as they enter/exit the top rankings
7. Flexible Styling: Configurable colors, fonts, units, and decimal formatting.

Technical Implementation:
1. Extends cv.sub_plot base class for seamless integration.
2. Leverages existing frame update logic and animation system.
3. Supports standard SJVisualizer configuration structure.
4. Includes responsive design that adapts to monitor dimensions.
5. Handles image loading from assets/ and asset1/ directories.

Attachments:
1. NewAxis - [ColumnAxis.py](https://github.com/user-attachments/files/23752893/ColumnAxis.py)
2. Example Usage via Jupyter Notebook file - [International Oil Companies Earnings Q1 2024 - Q2 2025.ipynb](https://github.com/user-attachments/files/23753234/International.Oil.Companies.Earnings.Q1.2024.-.Q2.2025.ipynb)
3. Output Video - https://github.com/user-attachments/assets/c4dab0e4-f898-4bed-85c4-1b5cb81cc848
4. data - [Earnings.xlsx](https://github.com/user-attachments/files/23753341/Earnings.xlsx)
5. assets: Company's CEOs (positioned above bars) & Background Image
<img width="2016" height="2016" alt="Chevron" src="https://github.com/user-attachments/assets/3af6e21b-9cf1-4972-af02-0c91d9ed7198" /> <img width="1200" height="1600" alt="Earnings" src="https://github.com/user-attachments/assets/0671fe87-a361-4340-8b81-145b2509728d" /> <img width="2016" height="2016" alt="ExxonMobil" src="https://github.com/user-attachments/assets/d00d17ca-3ba4-4b43-b0c4-ab20c54cb45f" /> <img width="2016" height="2016" alt="Shell" src="https://github.com/user-attachments/assets/1a5c011f-a6de-4a5f-9d96-0c7e36cb6870" /> <img width="2016" height="2016" alt="TotalEnergies" src="https://github.com/user-attachments/assets/aa759381-0d14-408e-b445-d573e805f524" />
6. asset1: Company's Logo (positioned inside bars)
<img width="2016" height="2016" alt="Chevron" src="https://github.com/user-attachments/assets/f2066465-9545-42ba-95be-73b361de803f" /> <img width="2016" height="2016" alt="ExxonMobil" src="https://github.com/user-attachments/assets/dd8bee68-c27a-45e5-9bdf-0cba615cb61f" /> <img width="2016" height="2016" alt="Shell" src="https://github.com/user-attachments/assets/7d22bb2e-30dc-45d6-99df-998fcc9dcbb3" /> <img width="2016" height="2016" alt="TotalEnergies" src="https://github.com/user-attachments/assets/fe0704a6-a7c3-4d3e-b8b0-2c8faf063605" />

File Structure:
1. column_race class: Main visualization controller.
6. column class: Individual bar element with animation capabilities
9. Compatible with pandas DataFrames for data input.

The implementation has been tested with Python 3.11 and includes example usage for rapid prototyping and testing.

AniekpongEffiong committed:
Create ColumnRace.py

This PR introduces a new column_race chart class for creating animated bar chart race visualizations, designed to work seamlessly with the SJVisualizer framework.

Key Features:
1. Animated Column Transitions: Smoothly animates bar height changes and positional movements as data rankings evolve over time.
2. Top-N Ranking Display: Automatically sorts and displays the top N data entries at each frame.
3. Dual Image Support: Internal column images (positioned inside bars) & External floating images (positioned above bars).
4. Negative Value Handling: Supports negative values with distinct red coloring and inverted bar direction.
5. Customizable Axes: Integrated with ColumnAxis.py for professional gridlines and scaling.
6. Dynamic Data Updates: Efficiently adds/removes columns as they enter/exit the top rankings
7. Flexible Styling: Configurable colors, fonts, units, and decimal formatting.

Technical Implementation:
1. Extends cv.sub_plot base class for seamless integration.
2. Leverages existing frame update logic and animation system.
3. Supports standard SJVisualizer configuration structure.
4. Includes responsive design that adapts to monitor dimensions.
5. Handles image loading from assets/ and asset1/ directories.

Attachments:
1. NewAxis - [ColumnAxis.py](https://github.com/user-attachments/files/23752893/ColumnAxis.py)
2. Example Usage via Jupyter Notebook file - [International Oil Companies Earnings Q1 2024 - Q2 2025.ipynb](https://github.com/user-attachments/files/23753234/International.Oil.Companies.Earnings.Q1.2024.-.Q2.2025.ipynb)
3. Output Video - https://github.com/user-attachments/assets/c4dab0e4-f898-4bed-85c4-1b5cb81cc848
4. data - [Earnings.xlsx](https://github.com/user-attachments/files/23753341/Earnings.xlsx)
5. assets:
<img width="2016" height="2016" alt="Chevron" src="https://github.com/user-attachments/assets/3af6e21b-9cf1-4972-af02-0c91d9ed7198" />
<img width="1200" height="1600" alt="Earnings" src="https://github.com/user-attachments/assets/0671fe87-a361-4340-8b81-145b2509728d" />
<img width="2016" height="2016" alt="ExxonMobil" src="https://github.com/user-attachments/assets/d00d17ca-3ba4-4b43-b0c4-ab20c54cb45f" />
<img width="2016" height="2016" alt="Shell" src="https://github.com/user-attachments/assets/1a5c011f-a6de-4a5f-9d96-0c7e36cb6870" />
<img width="2016" height="2016" alt="TotalEnergies" src="https://github.com/user-attachments/assets/aa759381-0d14-408e-b445-d573e805f524" />
6. asset1:
<img width="2016" height="2016" alt="Chevron" src="https://github.com/user-attachments/assets/f2066465-9545-42ba-95be-73b361de803f" />
<img width="2016" height="2016" alt="ExxonMobil" src="https://github.com/user-attachments/assets/dd8bee68-c27a-45e5-9bdf-0cba615cb61f" />
<img width="2016" height="2016" alt="Shell" src="https://github.com/user-attachments/assets/7d22bb2e-30dc-45d6-99df-998fcc9dcbb3" />
<img width="2016" height="2016" alt="TotalEnergies" src="https://github.com/user-attachments/assets/fe0704a6-a7c3-4d3e-b8b0-2c8faf063605" />

File Structure:
1. column_race class: Main visualization controller.
6. column class: Individual bar element with animation capabilities
7. Compatible with pandas DataFrames for data input.

The implementation has been tested with Python 3.11 and includes example usage for rapid prototyping and testing.

SjoerdTilmans commented:
Hi @AniekpongEffiong ,

Here’s a cleaned-up version that keeps your tone appreciative, but reads a bit more polished and clear:

Thanks for opening this pull request — and for the contribution!

A couple of questions / points before we merge:

What led you to create a new ColumnAxis file rather than reusing the existing Axis implementation?

I’ve left a few suggestions in the review that would be good to address first:

The text sizes for the category labels (e.g., “Chrome”) and the values (e.g., “11.08”) don’t currently scale with the configured font_size.

When categories fall outside the top X, their labels still get drawn in an odd (and visible) location. Could those labels either be moved off-screen or omitted entirely?

Apologies for the delay in getting this reviewed. I really appreciate the effort — this chart type fills an obvious gap in the library. Happy to collaborate to get it into great shape 🙂

<img width="2292" height="1902" alt="image" src="https://github.com/user-attachments/assets/db69004e-ddca-430e-80c9-f072947884af" />

SjoerdTilmans committed:
Added some test cases

AniekpongEffiong commented:
Hi @SjoerdTilmans,

Thank you again for taking the time to review this! I have carefully gone through all the issues you raised and addressed them in this updated commit (Updated ColumnRace.py file is attached).

[ColumnRace.py](https://github.com/user-attachments/files/25391043/ColumnRace.py) - Updated

Below is a detailed response to each of your questions:
**1. Text Sizes for Category Labels and Value Labels**
This issue has been fixed. The ColumnRace chart now properly reads the font_scale parameter from the user’s configuration, allowing text to scale correctly. To adjust the font size, users simply need to include the following in their Jupyter notebook when creating a ColumnRace instance:

font_scale = 1   or any preferred scaling factor

(This ensures that both category and value labels are rendered at the desired size throughout the animation).

**2. Labels for Categories Outside the Top-X**

The problem with “odd” visible labels has also been resolved. Previously, these labels appeared because label objects were repositioned but not hidden. In the updated implementation:

- Out-of-range label objects are now either deleted or moved completely off-screen.
- None of these labels render visually.
- Only the relevant top-X labels are displayed at each frame.

This guarantees a clean and accurate visualization for all frames.

**3. Why a New ColumnAxis.py Instead of Using the Existing Axis.py?**

I introduced a separate ColumnAxis class because ColumnRace has specific requirements that differ from the standard charts in the Library:

- **Dynamic Updates:** Because ColumnRace updates the axis on every frame with per-bar synchronization and gridline blending, a dedicated class ensures these features do not interfere with other chart types.
- **Spacing & Negative Values:** ColumnAxis supports variable bar widths, category-centered alignment, optional negative values, and gridlines behind moving bars. The original Axis does not provide this functionality.
- **Tick Layout:** Since ColumnRace is categorical and animated, it requires one tick per bar, precise text placement, and stable anchoring during reordering. The existing Axis, designed for continuous axes with “nice ticks”, isn’t compatible with this approach.
- **Category Labels:** The animation requires dynamically created, scalable labels integrated with column colors. The original Axis relies on fixed ticks that cannot adjust per frame.

<img width="823" height="688" alt="ColumnRace" src="https://github.com/user-attachments/assets/fada6862-e412-4510-af24-54137e3dfa7c" />
